src/solver/conv_asm_implicit_gemm_wrw_v4r1_dynamic.cpp

```diff
@@ -0,0 +1,406 @@
+
+    if((k % 128 == 0) && (GemmN % 128 == 0))
+        GemmKPerBlock = 16;
+    else
+        GemmKPerBlock = 4;
+    int gemmk_groups  = GetImplicitGemmWrwV4R1DynamicGemmkGroups(ctx.conv_problem, GemmKPerBlock);
+
+    result.workspce_sz = GetWorkspaceSize(ctx);
+
+    kernel.kernel_file = "igemm_v4r1_wrw_dynamic.s";
+    kernel.kernel_name = kernel_name;
+    kernel.g_wk.clear();
+    /* Note here, for API like hipHccModuleLaunchKernel(), hipExtModuleLaunchKernel()
+    * grid dims is in unit of work item.
+    * But for api like hipModuleLaunchKernel(), grid dim is in unit of block.
+    */
+    kernel.g_wk.push_back(grid_size * block_size);
+    kernel.g_wk.push_back(1);
+    kernel.g_wk.push_back(1);
+    kernel.l_wk.clear();
+    kernel.l_wk.push_back(block_size);
+    kernel.l_wk.push_back(1);
+    kernel.l_wk.push_back(1);
+
+    GenerateClangDefsym(options, "ROCM_METADATA_VERSION", ctx.rmv.UseV3() ? 5 : 4);
+
+    kernel.comp_options = options.str();
+
+    MIOPEN_LOG_I2(kernel.kernel_file + ":" + kernel.kernel_name);
+
+    result.construction_params.push_back(kernel);
+
+    if(gemmk_groups > 0)
+    {
+        KernelInfo kernel_reduction;
+        int reduction_per_thread     = 4;
+        kernel_reduction.kernel_file = "wrw_reduction_hip.cpp";
+        kernel_reduction.kernel_name = "wrw_reduction_hip";
+        kernel_reduction.g_wk.clear();
+        int block_size_reduction = 256;
+        int grid_size_redcution  = ctx.n_outputs * ctx.n_inputs * ctx.kernel_size_h *
+                                  ctx.kernel_size_w / (reduction_per_thread * block_size_reduction);
+        kernel_reduction.g_wk.push_back(grid_size_redcution * block_size_reduction);
+        kernel_reduction.g_wk.push_back(1);
+        kernel_reduction.g_wk.push_back(1);
+        kernel_reduction.l_wk.clear();
+        kernel_reduction.l_wk.push_back(block_size_reduction);
+        kernel_reduction.l_wk.push_back(1);
+        kernel_reduction.l_wk.push_back(1);
+
+        kernel_reduction.comp_options =
+            std::string("-Wno-old-style-cast ") + std::string("-Wno-cast-align");
+
+        result.construction_params.push_back(kernel_reduction);
+    }
+
+    const auto& conv_problem = ctx.conv_problem;
+
+    result.invoker_factory = [conv_problem](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle, const boost::any& primitive_parameters) {
+            const auto data_ctx = boost::any_cast<conv::WrWInvokeParams>(primitive_parameters);
+            const auto& tensors = data_ctx.tensors;
+            MIOPEN_LOG_I("wrw workspace size: " << data_ctx.workSpaceSize);
+            const auto& workSpace = data_ctx.workSpace;
+            std::vector<KernelInvoke> ks;
+            std::transform(kernels.begin(),
+                           kernels.end(),
+                           std::back_inserter(ks),
+                           [&](const Kernel& k_wrw) { return handle.Run(k_wrw); });
+            float elapsed = 0;
```

Comment:
To accurately compute elapsed time, you should have accumulated kernel time inside transform function.
{
handle.Run(k_wrw)
elapsed += handle.GetKernelTime();
}

Otherwise, would you not accumulating only reduction time?

Reply:
Also, please double check at dynamic bwd and fwd kernels, to make sure that we are accumulating elapsed time properly. Hopefully, we didn't miss it during code review.

Reply:
@TejashShah I think that this API handle.Run() only do collect kernel from convolution handle, and it does not actually run the kernel code in GPU device. If I collect  handle.GetKernelTime(); here, it would return zero. Am I right?
The reduction time is accumulated in function `CallImplicitGemmWrwDynamic(handle, conv_problem, tensors.x, tensors.dy, tensors.dw, workSpace, ks)`

Reply:
@shaojiewang You are saying that the actual kernel calls happens inside CallImplicitGemmWrwDynamic at the following lines of code. Then, you are right that the elapsed time returned from this function will capture all kernel invocations time.

```
    kernel(opArgs);

    if(handle.IsProfilingEnabled())
        elapsed += handle.GetKernelTime();
```